File: Python Codes/7_Best_Time_to_Buy_and_Sell_Stock.py
class Solution:
    def maxProfit(self, prices):
        if not prices:
            return 0

        ans = 0
        min_price = prices[0]

        for price in prices:
            ans = max(ans, price - min_price)
            min_price = min(min_price, price)

        return ans

if __name__ == "__main__":
    sol = Solution()
    prices = [7, 2, 5, 3, 6, 4] 

    print("Max Profit:", sol.maxProfit(prices))

File: CPP Codes/7_Best_Time_to_Buy_and_Sell_Stock.cpp
#include <iostream>
#include <vector>
#include <algorithm>

class Solution {
public:
    int maxProfit(std::vector<int>& prices) {
        if (prices.empty()) {
            return 0;
        }

        int ans = 0;
        int min = prices[0];

        for (int price : prices) {
            ans = std::max(ans, price - min);
            min = std::min(min, price);
        }

        return ans;
    }
};

int main() {
    Solution sol;
    std::vector<int> prices = {7, 1, 5, 3, 6, 4}; 
    std::cout << "Max Profit: " << sol.maxProfit(prices) << std::endl;
    return 0;
}
